NuRadioReco/examples/AliasPhasedArray/Noise_trigger_rate/noise_trigger_rate.py

```diff
@@ -0,0 +1,303 @@
+        Butterworth filter with critical frequencies:
+        (n-1) * adc_sampling_frequency/2 + bandwidth_edge
+        and
+        n * adc_sampling_frequency/2 - bandwidth_edge
+    bandwidth_edge: float
+        Frequency interval used for filtering the chosen Nyquist zone.
+        See above
+
+    Returns
+    -------
+    digital_trace: array of floats
+        Digitised voltage trace
+    adc_sampling_frequency: float
+        ADC sampling frequency for the channel
+    """
+
+    times = np.arange(len(trace)) / adc_sampling_frequency
+
+    adc_time_delay = 0
+
+    if random_clock_offset:
+        clock_offset = np.random.uniform(-1, 1)
+        adc_time_delay += clock_offset / adc_sampling_frequency
+
+    # Choosing Nyquist zone
+    if nyquist_zone is not None:
+
+        if nyquist_zone < 1:
+            error_msg = "Nyquist zone is less than one. Exiting."
+            raise ValueError(error_msg)
+        if not isinstance(nyquist_zone, int):
+            try:
+                nyquist_zone = int(nyquist_zone)
+            except:
+                error_msg = "Could not convert nyquist_zone to integer. Exiting."
+                raise ValueError(error_msg)
+
+        passband = ( (nyquist_zone-1) * adc_sampling_frequency/2 + bandwidth_edge,
+                     nyquist_zone * adc_sampling_frequency/2 - bandwidth_edge )
+        filtered_trace = butterworth_filter_trace(trace, adc_sampling_frequency,
+                                                  passband)
+
+    # Random clock offset
+    delayed_times = times + adc_time_delay
+    interpolate_trace = interp1d(times, trace, kind='quadratic',
+                                 fill_value='extrapolate')
+
+    delayed_trace = interpolate_trace(delayed_times)
+
+    interpolate_delayed_trace = interp1d(times, delayed_trace, kind='quadratic',
+                                         fill_value='extrapolate')
+
+    # Downsampling to ADC frequency
+    new_n_samples = int( (adc_sampling_frequency / input_sampling_frequency) * len(delayed_trace) )
+    resampled_times = np.linspace(0, new_n_samples/adc_sampling_frequency, new_n_samples)
+    resampled_trace = interpolate_delayed_trace(resampled_times)
+
+    # Digitisation
+    if adc_type == 'perfect_floor_comparator':
+        digital_trace = perfect_floor_comparator(resampled_trace, adc_n_bits,
+                                                 adc_ref_voltage, output)
+    elif adc_type == 'perfect_ceiling_comparator':
+        digital_trace = perfect_ceiling_comparator(resampled_trace, adc_n_bits,
+                                                   adc_ref_voltage, output)
+
+    # Upsampling with an FIR filter (if desired)
+    if upsampling_factor is not None:
+        if (upsampling_factor >= 2):
+            upsampling_factor = int(upsampling_factor)
+            upsampled_trace = upsampling_fir(digital_trace, adc_sampling_frequency,
+                                             int_factor=upsampling_factor, ntaps=2**7)
```

Comment:
ntaps probably needs to be a function of the upsampling_factor? (or configurable)

Reply:
I guess I commented on the wrong commit, but same principle stands if you want this to be a little more general :).  Basically the reason is that the number of taps needed will scale with the upsampling factor because the taps are calculated on the upsampled waveform.

Reply:
I've used 16 because that's what you told me. Is 16 fine for an upsampling of 4 or less? After that, how should I scale?

Reply:
If you've done tests and showed that it works well for that window length then that's probably fine!  However, it probably needs to scale with the upsampling factor (unless firwin automatically scales it or something).  Maybe it's a moot point if higher upsampling factors are never used?  If the window length is ever shorter than the number of zeroes stuffed between samples, though, that will cause weird problems.  Maybe upsampling_factor * 4 isn't a terrible rule of thumb?  You can also just sidestep it by making it an argument, since it could be an important parameter to vary in a study.

Reply:
I have implemented your rule of thumb. The number of taps can be specified now using the detector configuration, in the same way as ADC sampling frequency or reference voltage.